src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,262 @@
+// caller has to send the request to service control.
+// Otherwise, returns OK and cached response.
+::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse* response) {
+  if (request.service_name() != service_name_) {
+    return Status(Code::INVALID_ARGUMENT,
+                  (string("Invalid service name: ") + request.service_name() +
+                   string(" Expecting: ") + service_name_));
+  }
+
+  if (!request.has_allocate_operation()) {
+    return Status(Code::INVALID_ARGUMENT,
+                  "allocate operation field is required.");
+  }
+
+  if (!cache_) {
+    // By returning NO_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+  if (!lookup.Found()) {
+    // To prevent other connections from sending their own AllocateQuoteRequest,
+    // insert a temporary positive response to the cache. Requests from other
+    // connections will be aggregated to this temporary element until the
+    // response for the actual request arrives.
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse temp_response;
+    CacheElem* cache_elem = new CacheElem(temp_response);
+    cache_elem->set_signature(request_signature);
+    cache_->Insert(request_signature, cache_elem, 1);
+
+    // By returning NOT_FOUND, caller will send AllocateQuotaRequest
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  // Aggregate tokens if the cached response is positive
+  if (lookup.value()->quota_response().allocate_errors_size() == 0) {
+    lookup.value()->Aggregate(request);
+  }
+
+  *response = lookup.value()->quota_response();
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// Caches a response from a remote Service Controller AllocateQuota call.
+::google::protobuf::util::Status QuotaAggregatorImpl::CacheResponse(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    const ::google::api::servicecontrol::v1::AllocateQuotaResponse& response) {
+  if (!cache_) {
+    return ::google::protobuf::util::Status::OK;
+  }
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  CacheElem* cache_elem = nullptr;
+
+  if (response.allocate_errors_size() == 0) {
+    // If the current response is positive and the signature is already in the
```

Comment:
When the response come,  cache should have its entry since it was inserted when sending the request.  
If the entry is gone,  we don't need to do anything.

so remove lines from 163 to 173